# Sesión 08 · Agrupa, resume y devuelve contexto

Badge 2 · Analista con Pandas

Distingue agregación, transformación y selección de grupos, con controles de conservación.

150 minutos de clase y 45–60 minutos de tarea. Ejecuta en orden.

## Agrupar para responder una pregunta

groupby divide registros según una o varias claves. Una agregación resume cada grupo con sum, mean, count u otra operación. Elige el resumen según la pregunta.

Sumar cantidades devuelve unidades; contar filas devuelve registros. Son indicadores diferentes, aunque a veces tengan el mismo valor.

In [1]:
import pandas as pd

ventas = pd.DataFrame({"Producto": ["Mouse", "Mouse", "Teclado"],
                      "Cantidad": [3, 1, 2], "Ingreso": [47.97, 15.99, 51.0]})
resumen = ventas.groupby("Producto", as_index=False).agg(
    unidades=("Cantidad", "sum"),
    ingreso=("Ingreso", "sum"),
    registros=("Producto", "size")
)
print(resumen)

  Producto  unidades  ingreso  registros
0    Mouse         4    63.96          2
1  Teclado         2    51.00          1


**Qué observar:** Mouse: 4 unidades, 63.96 de ingreso y 2 registros. Teclado: 2, 51.00 y 1.

**Recuerda:** Un promedio de precios por registro no es lo mismo que el precio promedio ponderado por unidades.

## transform: una estadística por fila

agg reduce cada grupo a un resumen; transform devuelve un resultado alineado con las filas originales. Sirve para añadir el total del grupo y calcular la participación de cada registro.

Prefiere transform y operaciones vectorizadas para cálculos conocidos. apply es flexible, pero puede ser menos claro y eficiente. Define cómo tratar un grupo cuyo total sea cero antes de dividir.

In [2]:
import pandas as pd
v=pd.DataFrame({"Producto":["A","A","B"],"Ingreso":[10,30,20]})
v["TotalGrupo"]=v.groupby("Producto")["Ingreso"].transform("sum")
v["Participacion"]=v.Ingreso/v.TotalGrupo
print(v)

  Producto  Ingreso  TotalGrupo  Participacion
0        A       10          40           0.25
1        A       30          40           0.75
2        B       20          20           1.00


**Qué observar:** A: participaciones 0.25 y 0.75; B: 1.0. Se conservan 3 filas.

**Recuerda:** transform añade contexto de grupo sin perder la granularidad original.

## filter y claves ausentes

filter conserva grupos completos según una condición. No equivale a filtrar filas por su valor individual: un grupo puede conservar todas sus filas aunque algunas sean pequeñas.

groupby omite claves faltantes por defecto. Usa dropna=False si quieres incluir un grupo para ellas, y compara totales. size cuenta filas; count cuenta valores no nulos en la columna seleccionada.

In [3]:
import pandas as pd
v=pd.DataFrame({"Grupo":["A","A","B",None],"Valor":[10,30,20,5]})
print(v.groupby("Grupo",dropna=False).Valor.sum())
print(v.groupby("Grupo",dropna=False).filter(lambda g: len(g)>=2))
assert v.groupby("Grupo",dropna=False).Valor.sum().sum()==v.Valor.sum()

Grupo
A      40
B      20
NaN     5
Name: Valor, dtype: int64
  Grupo  Valor
0     A     10
1     A     30


**Qué observar:** Total conservado 65. filter conserva las dos filas de A.

**Recuerda:** Una suma por grupos puede perder observaciones si hay claves faltantes y no revisas dropna.

## Reto · Total por producto

Agrupa A, B, A con ingresos 100, 200, 50. Ordena los totales de mayor a menor.

**Criterio:** B: 200; A: 150.

In [ ]:
# Tu solución aquí.

## Reto · Porcentaje dentro del producto

Para A=[10,30] y B=[20], añade participaciones por fila.

**Criterio:** 25 %, 75 %, 100 %.

In [ ]:
# Tu solución aquí.

## Tarea · Indicadores por producto

Usa ventas.csv, calcula Ingreso y mantén una copia original.

1. Calcula unidades, ingresos y registros con agg.
2. Añade participación de cada venta en el ingreso del producto con transform.
3. Selecciona grupos con al menos dos registros.
4. Introduce una clave faltante y demuestra cómo dropna afecta la conservación del total.

In [ ]:
# Desarrolla tu tarea y explica los resultados.